Ask for local secret param when collecting user inputs

elvisun opened the pull request:
### Description
Ask for local secret param, then write it to extensions manifest.

## When creating a secret, user will be asked how they want to handle the secret.
![Screen Shot 2022-03-16 at 10 38 25 AM](https://user-images.githubusercontent.com/6955348/158616059-24c1d558-f15b-4830-be73-543edc231597.png)

## New prompt for entering local secret.
![Screen Shot 2022-03-16 at 10 38 47 AM](https://user-images.githubusercontent.com/6955348/158616061-9d24105f-aacc-4d90-a64b-28ca9ff8f675.png)

## Install flow for only using secret manager
![Screen Shot 2022-03-16 at 10 37 58 AM](https://user-images.githubusercontent.com/6955348/158616052-878b9a99-001f-4766-ad0e-d354cee06da2.png)

## Install flow for using both secret manager and local
![Screen Shot 2022-03-16 at 10 38 58 AM](https://user-images.githubusercontent.com/6955348/158616063-f599e9ea-4935-46be-ae59-e8fdc6ee50a3.png)

## Reconfigure
![Screen Shot 2022-03-16 at 11 18 31 AM](https://user-images.githubusercontent.com/6955348/158624821-a06d0b19-3c7a-420e-9c1d-9eeadcca056c.png)


#### Alternatives Considered
- Asking a yes or no questions on configuring a local secret after setting a cloud secret. This doesn't work because users might want to **only** configure a local secret.

### Scenarios Tested

Manually tested install configure update

joehan commented:
Gonna start reviewing this now - however, it looks like the git history is funky, probably because this includes all the individual commits from #4292 that got squashed into master. If it doesn't create too much of a headache, could you rebase this onto master?

joehan commented on src/extensions/askUserForParam.ts:
```diff
@@ -150,16 +157,42 @@ export async function askForParam(args: {
+            ],
+          });
+        }
+        if (secretLocations.includes(SecretCreationAction.CLOUD.toString())) {
```
Optional: This chunk from l187 to l193 could be good to separate into its own function, for readability.

joehan commented on src/deploy/extensions/planner.ts:
```diff
@@ -6,7 +6,30 @@ import { FirebaseError } from "../../error";
+ * So far this is only used for writing to the manifest, but in the future
+ * we want to read manifest into this interface.
+ */
+export interface ManifestInstanceSpec {
```
I think creating this type split is reasonable for now, but I'd prefer to join these into a single type in the future. They have a ton of overlap (and InstanceSpec has some useful caching functions to fetch extension & extensionVersion), and the only difference is params, so I'd rather we move the type split down to just params. Maybe something like:
```
export type ParamValue = string | ParamBindingOptions
export function resolveParamValues(p: ParamValue) {
// Either resolves the ParamBindingOption or returns the string value
}
```

elvisun replied:
I explored in that direction but not exactly a union type, I decided to not touch `InstanceSpec` because it's used in too many places, and it'd be hard to untangle..

The reason why I ended up introducing another type is because I thought conceptually it made sense. `ManifestInstanceSpec` is the bare minimum spec that we pull from the manifest. But after hydration with ExtensionVersion, Extension, and collapsing the params , it turns into `DeploymentInstanceSpec`, something we use for deployment/emulator. So in the long term I'm leaning towards keeping both types but maybe make them more clearly separated? wdyt?

joehan commented on src/extensions/manifest.ts:
```diff
@@ -56,6 +57,44 @@ export async function writeToManifest(
+    }
+
+    let writeBuffer: Record<string, string> = {};
+    for (const paramSpec of spec.paramSpecs.filter((p) => p.type === ParamType.SECRET)) {
```
Nit: I think this code ends up cleaner if we also filter for only params that have a local override- ie:
```
const locallyOverridenSecretParams = spec.paramsSpecs.filter((p) => p.type === ParamType.Secret && spec.params[p.param].local)
```
```

elvisun replied:
good suggestion, done

joehan commented on src/extensions/askUserForParam.ts:
```diff
@@ -6,15 +6,20 @@ const { marked } = require("marked");
 import { ParamBindingOptions } from "./paramHelper";
 
+enum SecretCreationAction {
+  CLOUD = 1,
```
I think i might just be not understanding some nuance of TS enums, but why did you do this `CLOUD = 1, LOCAL,`? Does the `=1` set a default or something?

elvisun replied:
it's just always better to start enum at 1 :)

Don't think this made it into the ts style but we have a firebase style guide that recommended this:
go/fireconsole-styleguide#always-start-enums-with-1

joehan reviewed (commented):
Took a pass through with some comments - will come back once tests are passing too

elvisun commented:
> Gonna start reviewing this now - however, it looks like the git history is funky, probably because this includes all the individual commits from #4292 that got squashed into master. If it doesn't create too much of a headache, could you rebase this onto master?

I tried to rebase but didn't look like a good idea... I'll leave this as is so we don't risk missing this for the bug bash. Since we are going to squash merge I think it's ok?

Fixed the tests and addressed the comments. I'll come back after dinner to add some unit tests and do some more manual testing and polishing

joehan reviewed (approved):
This LGTM as is - we should add some unit test for writeLocalSecrets, but I think it is OK to merge this in time for the bug bash and add them afterwards. No need to work late into the night for this one :)

elvisun commented:
> This LGTM as is - we should add some unit test for writeLocalSecrets, but I think it is OK to merge this in time for the bug bash and add them afterwards. No need to work late into the night for this one :)

Thanks for the review! no worries, wanted to do another pass before the bugbash :) Just added some unit tests, going to merge this so I can send you some other small fixes
